pkg/middleend/qos.go

```diff
@@ -65,30 +49,44 @@ func (s *Server) DeleteQosVolume(_ context.Context, in *pb.DeleteQosVolumeReques
 		log.Printf("error: %v", err)
 		return nil, err
 	}
-	params := spdk.BdevQoSParams{
-		Name:           qosVolume.VolumeId.Value,
-		RwIosPerSec:    0,
-		RwMbytesPerSec: 0,
-		RMbytesPerSec:  0,
-		WMbytesPerSec:  0,
-	}
-	var result spdk.BdevQoSResult
-	err := s.rpc.Call("bdev_set_qos_limit", &params, &result)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, spdk.ErrFailedSpdkCall
-	}
-	log.Printf("Received from SPDK: %v", result)
-	if !result {
-		msg := fmt.Sprintf("Could not clear QoS limit for : %s", in.Name)
-		log.Print(msg)
-		return nil, spdk.ErrUnexpectedSpdkCallResult
+
+	if err := s.cleanMaxLimit(qosVolume.VolumeId.Value); err != nil {
+		return nil, err
 	}
 
 	delete(s.volumes.qosVolumes, in.Name)
 	return &emptypb.Empty{}, nil
 }
 
+// UpdateQosVolume updates a QoS volume
+func (s *Server) UpdateQosVolume(_ context.Context, in *pb.UpdateQosVolumeRequest) (*pb.QosVolume, error) {
+	log.Printf("UpdateQosVolume: Received from client: %v", in)
+	if err := s.verifyQosVolume(in.QosVolume); err != nil {
+		log.Println("error:", err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	qosVolumeID := in.QosVolume.QosVolumeId.Value
+	volume, ok := s.volumes.qosVolumes[qosVolumeID]
+	if !ok {
+		log.Printf("Non-existing QoS volume with id %v", qosVolumeID)
+		return nil, status.Errorf(codes.NotFound, "volume_id %v does not exist", qosVolumeID)
+	}
+
+	if volume.VolumeId.Value != in.QosVolume.VolumeId.Value {
```

Comment:
if this correct or error use case ?

Reply:
simpler to start is to forbid this reassignment first

Reply:
updated to return an error